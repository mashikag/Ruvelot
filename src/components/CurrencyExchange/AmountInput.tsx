import React, { ChangeEvent, KeyboardEvent, ReactElement, useCallback } from 'react'
import styled from 'styled-components'
import { safeAmount } from '../../utils/currencyExchangeUtils'

export interface ExchangeAmountInputProps {
  amount: number | null
  onChange: (amount: number | null) => void
}

function AmountInput(
  { amount, onChange }: ExchangeAmountInputProps
): ReactElement {

  const safeOnAmountChange =
    ({ target }: ChangeEvent<HTMLInputElement>) => {
      if (target.value === '') {
        onChange(null)
      } else {
        onChange(safeAmount(target.value))
      }
    }

  const safeOnAmountChangeCallback = useCallback(
    safeOnAmountChange,
    [onChange]
  )

  return <StyledAmountInput
    onChange={safeOnAmountChangeCallback}
    onKeyPressCapture={preventNotSupportedKeys}
    min={0}
    max={99999999}
    step='0.01'
    type='number'
    value={amount === null ? '' : safeAmount(amount)}
  />
}

export default AmountInput

const StyledAmountInput =
  styled.input({
    display: 'block',
    fontSize: '3em',
    width: '100%',
    textAlign: 'right',
    border: 'none',
    ':invalid': {
      color: '#be0000'
    }
  })

const otherSupportedAmountInputKey = {
  backspace: 8,
  fullStop: 46,
}

function preventNotSupportedKeys(evt: KeyboardEvent<HTMLInputElement>): void {
  const keyCode = (evt.which) ? evt.which : evt.keyCode
  if (!isKeyCodeSupported(keyCode)) {
    evt.stopPropagation()
    evt.preventDefault()
  }
}

function isDigitKeyCode(keyCode: number) {
  return keyCode >= 48 && keyCode <= 57
}

function isKeyCodeSupported(keyCode: number): boolean {
  if (isDigitKeyCode(keyCode)) {
    return true
  }
  return Object.values(otherSupportedAmountInputKey).includes(keyCode)
}