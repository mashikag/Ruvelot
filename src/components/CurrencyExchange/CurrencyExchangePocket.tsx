import React, { ChangeEventHandler, ReactElement, ChangeEvent } from 'react'
import styled from 'styled-components'
import { ExchangeRates, Currency } from '../../store/types'
import { safeAmount } from '../../utils/currencyExchangeUtils';

const CurrencySelector = styled.select<React.SelectHTMLAttributes<HTMLSelectElement>>({})
const Pocket = styled.div<React.HTMLAttributes<HTMLDivElement>>({})
const ExchangeAmountInput = styled.input<React.InputHTMLAttributes<HTMLInputElement>>({})

export interface CurrencyExchangePocketProps {
  currency: Currency
  amount: number | null
  onCurrencyChange: ChangeEventHandler<HTMLSelectElement>
  onAmountChange: (amount: number | null) => void
  rates: ExchangeRates
}

function createCurrencySelectorOptions(exchangeRates: ExchangeRates):
  JSX.Element[] {
  return Object.keys(exchangeRates).map<JSX.Element>(
    currency => <option key={currency} value={currency}>{currency}</option>
  )
}

function CurrencyExchangePocket(props: CurrencyExchangePocketProps):
  ReactElement<CurrencyExchangePocketProps> {

  const safeOnAmountChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
    if (target.value === '') {
      target.value = '0'
      props.onAmountChange(null)
    }
    props.onAmountChange(parseFloat(parseFloat(target.value).toFixed(2)))
  }

  return (
    <Pocket>
      <CurrencySelector onChange={props.onCurrencyChange} value={props.currency}>
        {createCurrencySelectorOptions(props.rates)}
      </CurrencySelector>
      <ExchangeAmountInput
        type='number'
        min='0'
        value={props.amount ? safeAmount(props.amount) : ''}
        onChange={safeOnAmountChange}
      />
    </Pocket>
  );
}

export default CurrencyExchangePocket
