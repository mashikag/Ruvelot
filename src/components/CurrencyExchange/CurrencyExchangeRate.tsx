import React, { ReactElement } from 'react'
import { Currency } from '../../store/types'

export interface CurrencyExchangeRateProps {
  fromCurrency: Currency
  toCurrency: Currency
  rate: number
}

function CurrencyExchangeRate(props: CurrencyExchangeRateProps):
  ReactElement<CurrencyExchangeRateProps> {
  return (
    <div>
      {`1 ${props.fromCurrency} = ${props.rate} ${props.toCurrency} `}
    </div>
  )
}

export default CurrencyExchangeRate