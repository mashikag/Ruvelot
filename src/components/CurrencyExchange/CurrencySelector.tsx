import React from 'react'

const CurrencySelector: React.SFC<{}> =
  () => <div></div>

export default CurrencySelector
