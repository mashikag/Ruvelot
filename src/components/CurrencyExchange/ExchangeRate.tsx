import React, { ReactElement } from 'react'
import styled from 'styled-components'
import { Currency, StoreState } from '../../store/types'
import { useSelector, shallowEqual } from 'react-redux'

export interface ExchangeRateState {
  fromCurrency: Currency
  toCurrency: Currency
  rate: number
}

function ExchangeRate(): ReactElement {
  const { fromCurrency, toCurrency, rate }: ExchangeRateState =
    useSelector(
      ({
        currencyExchange: {
          exchangeFromCurrency,
          exchangeToCurrency,
          rates
        }
      }: StoreState) => ({
        fromCurrency: exchangeFromCurrency,
        toCurrency: exchangeToCurrency,
        rate: rates[exchangeToCurrency] / rates[exchangeFromCurrency]
      }),
      shallowEqual
    )
  return (
    <ExchangeRateContainer>
      {`1 ${fromCurrency} = ${rate} ${toCurrency}`}
    </ExchangeRateContainer>
  )
}

export default ExchangeRate

const ExchangeRateContainer =
  styled.div({
    color: 'white',
    backgroundColor: '#0075eb',
    fontWeight: 'bold',
    padding: '20px',
    textAlign: 'center',
  })