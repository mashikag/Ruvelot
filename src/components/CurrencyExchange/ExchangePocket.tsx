import React, { ReactElement } from 'react'
import styled from 'styled-components'
import AmountInput from './AmountInput'
import CurrencySelect from '../CurrencySelect/CurrencySelect'
import { Currency, StoreState } from '../../store/types'
import { useSelector, shallowEqual } from 'react-redux'

export interface ExchangePocketProps {
  amount: number | null
  currency: Currency
  excludeCurrency: Currency
  label: string
  onAmountChange: (amount: number | null) => void
  onCurrencyChange: (value: Currency) => void
}

function ExchangePocket(
  { amount, currency, excludeCurrency, label, onAmountChange, onCurrencyChange }: ExchangePocketProps
): ReactElement {

  const currencies = useSelector<StoreState, Currency[]>(
    ({ currencyExchange: { rates } }) =>
      (Object.keys(rates) as Currency[])
        .filter(currency => currency !== excludeCurrency),
    shallowEqual
  )

  return (
    <Pocket>
      <Label>{label}</Label>
      <PocketInputsContainer>
        <CurrencySelect
          style={{ flex: 'auto' }}
          currencies={currencies}
          onCurrencyChange={onCurrencyChange}
          value={currency} />
        <AmountInput onChange={onAmountChange} amount={amount} />
      </PocketInputsContainer>
    </Pocket>
  );
}

export default ExchangePocket

const Label = styled.div({
  textAlign: 'left',
  color: 'gray',
  fontSize: '1.1em',
})

const Pocket = styled.div({
  display: 'flex',
  flexFlow: 'column nowrap',
  height: '100%',
  flex: 'auto',
  textAlign: 'right',
  border: '1px solid #eaeaea',
  borderRadius: '3px',
  padding: '20px'
})

const PocketInputsContainer = styled.div({
  display: 'flex',
  padding: '30px',
  flexFlow: 'row nowrap'
})
