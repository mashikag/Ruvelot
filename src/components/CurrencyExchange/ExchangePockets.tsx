import React, { ReactElement, useCallback } from 'react'
import styled from 'styled-components'
import ExchangePocket from './ExchangePocket'
import { Currency, StoreState } from '../../store/types';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { setExchangeFromCurrency, setExchangeToCurrency, setExchangeFromAmount, setExchangeToAmount } from '../../store';

const EXCHANGE_FROM_POCKET_LABEL = 'From'
const EXCHANGE_TO_POCKET_LABEL = 'To'

interface ExchangePocktsState {
  exchangeFromAmount: number | null
  exchangeToAmount: number | null
  exchangeFromCurrency: Currency
  exchangeToCurrency: Currency
}

function ExchangePockets(): ReactElement {

  const {
    exchangeFromAmount,
    exchangeToAmount,
    exchangeFromCurrency,
    exchangeToCurrency,
  } = useSelector<StoreState, ExchangePocktsState>(
    ({
      currencyExchange: {
        exchangeFromAmount,
        exchangeFromCurrency,
        exchangeToAmount,
        exchangeToCurrency
      }
    }) => ({
      exchangeFromAmount,
      exchangeToAmount,
      exchangeFromCurrency,
      exchangeToCurrency
    }),
    shallowEqual
  )

  const dispatch = useDispatch();

  const onExchangeFromCurrencyChange = useCallback(
    (currency: Currency) =>
      dispatch(setExchangeFromCurrency(currency)),
    [dispatch]
  )

  const onExchangeToCurrencyChange = useCallback(
    (currency: Currency) =>
      dispatch(setExchangeToCurrency(currency)),
    [dispatch]
  )

  const onExchangeFromAmountChange = useCallback(
    (amount: number | null) =>
      dispatch(setExchangeFromAmount(amount)),
    [dispatch]
  )

  const onExchangeToAmountChange = useCallback(
    (amount: number | null) =>
      dispatch(setExchangeToAmount(amount)),
    [dispatch]
  )

  return (
    <PocketsContainer>
      <ExchangePocket
        amount={exchangeFromAmount}
        currency={exchangeFromCurrency}
        excludeCurrency={exchangeToCurrency}
        label={EXCHANGE_FROM_POCKET_LABEL}
        onAmountChange={onExchangeFromAmountChange}
        onCurrencyChange={onExchangeFromCurrencyChange} />
      <ExchangePocket
        amount={exchangeToAmount}
        currency={exchangeToCurrency}
        excludeCurrency={exchangeFromCurrency}
        label={EXCHANGE_TO_POCKET_LABEL}
        onAmountChange={onExchangeToAmountChange}
        onCurrencyChange={onExchangeToCurrencyChange} />
    </PocketsContainer>
  )
}

export default ExchangePockets

const PocketsContainer = styled.div({
  display: 'flex'
}) 