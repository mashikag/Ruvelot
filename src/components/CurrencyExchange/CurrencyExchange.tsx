import React, { ReactElement } from 'react'
import styled from 'styled-components'
import ExchangePockets from './ExchangePockets'
import ExchangeRate from './ExchangeRate'
import SubmitExchangeButton from './SubmitExchangeButton'
import SwitchExchangeCurrencies from './SwitchExchangeCurrencies'

function CurrencyExchange(): ReactElement {
  return (
    <CurrencyExchangeContainer>
      <ExchangeRate />
      <SwitchExchangeCurrencies />
      <ExchangePockets />
      <SubmitExchangeButton />
    </CurrencyExchangeContainer>
  )
}

export default CurrencyExchange

const CurrencyExchangeContainer = styled.div({
  position: 'relative'
})