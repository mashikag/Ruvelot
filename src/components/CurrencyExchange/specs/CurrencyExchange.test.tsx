import React from 'react'
import { shallow } from 'enzyme'
import CurrencyExchange from '../CurrencyExchange'

describe('Currency Exchange Component', () => {
  it(`shallow snapshot`, () => {
    const component = shallow(<CurrencyExchange />)
    expect(component).toMatchSnapshot()
  })
})