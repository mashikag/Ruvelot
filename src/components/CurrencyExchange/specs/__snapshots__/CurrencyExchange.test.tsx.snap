// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Exchange Component shallow snapshot 1`] = `
<styled.div>
  <ExchangeRate />
  <SwitchExchangeCurrencies />
  <ExchangePockets />
  <SubmitExchangeButton />
</styled.div>
`;
